ledger/test-common/src/main/scala/com/digitalasset/platform/testing/TimeBoundObserver.scala

```diff
@@ -7,36 +7,28 @@ import com.daml.timer.Delayed
 import io.grpc.Context
 import io.grpc.stub.StreamObserver
 
+import scala.concurrent.ExecutionContext
 import scala.concurrent.duration.FiniteDuration
-import scala.concurrent.{ExecutionContext, Future, Promise}
 
-final class TimeBoundObserver[T](duration: FiniteDuration, maximumResults: Option[Int] = None)(
-    implicit executionContext: ExecutionContext)
-    extends StreamObserver[T] {
+final class TimeBoundObserver[A](
+    duration: FiniteDuration
+)(delegate: StreamObserver[A])(implicit executionContext: ExecutionContext)
+    extends StreamObserver[A] {
 
-  private val promise = Promise[Vector[T]]
-  private val buffer = Vector.newBuilder[T]
-  private var size = 0
-
-  Delayed.by(duration)(onCompleted())
-
-  def result: Future[Vector[T]] = promise.future
+  Delayed.by(duration)(synchronized {
+    onCompleted()
```

Comment:
What happens if the delegate cancels as part of its `onCompleted`?

Reply:
Then it'll be cancelled twice, which I don't think is really an issue. GRPC is quite lenient with repeated calls like this. (As any asynchronous protocol should be; you can never guarantee that you won't have a race somewhere.)

That said, I think I have made sure that `onCompleted` never cancels.